### ISSUE TITLE
MJS2020: More powersaving

### ISSUE THREAD

**Issue by matthijskooijman** [meetjestad/mjs_firmware#19]
For MJS2020, I now implemented basic powersaving based on the STM32 "Stop" mode. This mode stops all core clocks and some peripherals, puts the internal regulator in a low-power mode and is the lowest-power mode that preserves memory contents ("Standby mode" is even lower power, but greatly complicates programming because memory contents is lost, and the gain is limited - 0.45μA for stop, 0.29μA for standby).

There are still more things that can be done to improve power usage, both to improve the sleep power usage, as well as improve the overall power usage.

 - Checking other components on the board. In stop mode, the STM32 can be down to typ 0.45μA, so there's probably some more components drawing current that could maybe powered down.
 - Disabling more peripherals. AFAICS the STM32L0 library already powers down a lot of stuff during stop mode, but we should doublecheck if anything is left on (i.e. maybe it leaves the UART on because it is configured, or some part of USB?).
 - Putting (unused) pins in "analog input mode" (i.e. disabling the digital input circuitry), might save power when input signals are fluctuating (this might or might not be relevant while sleeping, but might save just a little power while running).
 - Sleep more often (i.e. also while waiting for the SPS030 to stabilize, or, more complicated because this requires Basicmac changes, while waiting for an RX window or for airtime to become available).
 - Lower CPU speed and/or peripheral clocks while running, which also allows running at a lower core voltage by reconfiguring the internal regulator. Especially while waiting for things when you cannot sleep (so where a lower speed does not mean running for longer), this can have significant savings (6-7mA for 32Mhz, only 190μA for 1Mhz).
 - Can the SPS030 maybe enable its fan without also measuring? We won't use the initial measurements anyway. Might not be possible and/or not have significant savings, though.

**Comment by matthijskooijman** [comment]
Wrt to stop mode and peripherals:
 - Before entering stop mode, the code (https://github.com/meetjestad/mjs_boards/blob/fac77ab752f6b1c881bc6a7c003304554e3c71d8/system/STM32L0xx/Source/stm32l0_system.c#L1500-L1629) disables some stuff (regulator in low-power mode, disable VREF) but not a lot.
 - Before entering stop mode, or checking peripherals, it checks for various locks which can prevent stop mode altogether, or keep things enabled.
 - `STM32L0_SYSTEM_LOCK_STOP` prevents stop mode from being entered at all. This lock is taken by USB (when attached/enabled) ADC (during conversion), DAC (when enabled), EEPROM/Flash (while writing), I²C (during a transaction), SDSPI (?), SPI (during transactions), UART (when wakup is enabled and maybe also during RX)
 - `STM32L0_SYSTEM_LOCK_REGULATOR` prevents the regulator from going into low-power mode. This lock is taken by UART (when wakup enabled and baud > 19200).
 - `STM32L0_SYSTEM_LOCK_VREFINT` prevents the voltage reference from being disabled. This lock is taken by the HSI48 clock (while enabled), which is enabled by the random generator (while generating a number), MCO (clock output, when enabled) and USB (when enabled).
 - Additionally, I think that peripherals that are enabled will keep one or more clocks running, so disabling peripherals (i.e. `SPI.end()` or `Serial1.end()` probably helps to lower power usage further.
 - I'm not sure how this works for USB, though. Lacking a VBUS sense pin, I suspect that USB is always "attached" (pullup enabled, ready to be enumerated) and might need a clock running, though the above suggests that even when detached but initialized. De-initializing USB seems to be more than just `Serial.end()`, though, since that just seems to disable the CDC serial part, not the USB core. Attach/detach seems to be possible using `USBDevice.attach()` / `USBDevice.detach()`, but there is no way to de-initialize, it seems.
